Updated GitHub actions versions

Zuzu-Typ opened the pull request:
## New GitHub actions versions
+ `actions/checkout` updated from v3 to v4 (see [actions/checkout](https://github.com/actions/checkout/releases))
+ `actions/setup-python` updated from v4 to v5 (see [actions/setup-python](https://github.com/actions/setup-python/releases))
+ `actions/upload-artifact` updated from v3 to v4 (see [actions/upload-artifact](https://github.com/actions/upload-artifact/releases))
+ `actions/download-artifact` updated from v3 to v4 (see [actions/download-artifact](https://github.com/actions/download-artifact/releases))

## Required changes
In `upload-artifact@v4` artifacts are immutable. Therefore uploading artifacts with the same name is no longer permitted. I adjusted the workflow according to [MIGRATION.md](https://github.com/actions/upload-artifact/blob/main/docs/MIGRATION.md).

johnnynunez commented:
and arm runners?

Zuzu-Typ commented:
> and arm runners?

Good idea. It would require some additional testing though.
That would potentially include the following architectures:
- MacOS apple silicon (arm64) (or universal2)
- Windows on Arm
- manylinux & musllinux aarch64, ppc64le, s390x and armv7l

It's not entirely clear though if all of those binaries can actually be built without modifying the code.

I can try, but it'll take a while.

johnnynunez commented:
> > and arm runners?
> 
> Good idea. It would require some additional testing though. That would potentially include the following architectures:
> 
> * MacOS apple silicon (arm64) (or universal2)
> * Windows on Arm
> * manylinux & musllinux aarch64, ppc64le, s390x and armv7l
> 
> It's not entirely clear though if all of those binaries can actually be built without modifying the code.
> 
> I can try, but it'll take a while.

yes, also you have arm runners now to test it:
https://github.blog/changelog/2025-01-16-linux-arm64-hosted-runners-now-available-for-free-in-public-repositories-public-preview/

digits, jetson thor, cuda arm laptops are coming...

Zuzu-Typ commented:
> yes, also you have arm runners now to test it: https://github.blog/changelog/2025-01-16-linux-arm64-hosted-runners-now-available-for-free-in-public-repositories-public-preview/

Nice, thanks for the heads-up!
I implemented the runners into a new workflow setup. I tested it and it seems to work quite well.
ppc64le and s390x will not be supported, because NumPy does not support them. I added the code anyway, but it's commented out until that changes (if ever).

johnnynunez commented:
@Zuzu-Typ FYI: also window arm is coming for Q2 2025 https://github.com/github/roadmap/issues/1098

johnnynunez commented:
Thanks for your runners, I really appreciate it from Nvidia (Digits, Thor and Cuda ARM laptops)

johnnynunez commented:
cc @mcfletch could you merge it ASAP?
![image](https://github.com/user-attachments/assets/06141edc-3fd1-4d6d-863c-1bef16cc1547)

johnnynunez commented:
One question,
Is it necessary mantain old python <3.9?
https://devguide.python.org/versions/
i mean in terms of cost, efficiency and security

I mean

Zuzu-Typ commented:
> One question, Is it necessary mantain old python <3.9? https://devguide.python.org/versions/ i mean in terms of cost, efficiency and security

Are you asking me or mcfletch?

I believe it isn't strictly necessary. 
I think it's generally a good idea to support older versions if there is no downside (no workarounds, no features that can't be used). Then the users of those platforms can still use the latest version with all bugfixes. Most packages (e,g, [numpy](https://pypi.org/project/numpy/2.0.0/#files)) don't do this though so one might run into compatibility problems.

I removed <=3.8 from the build script, because they might not be supported on all manylinux or musllinux images.

johnnynunez commented:
> > One question, Is it necessary mantain old python <3.9? https://devguide.python.org/versions/ i mean in terms of cost, efficiency and security
> 
> Are you asking me or mcfletch?
> 
> I believe it isn't strictly necessary. I think it's generally a good idea to support older versions if there is no downside (no workarounds, no features that can't be used). Then the users of those platforms can still use the latest version with all bugfixes. Most packages (e,g, [numpy](https://pypi.org/project/numpy/2.0.0/#files)) don't do this though so one might run into compatibility problems.
> 
> I removed <=3.8 from the build script, because they might not be supported on all manylinux or musllinux images.

I hope that @mcfletch merge ASAP
